common/display/src/main/scala/uk/ac/wellcome/display/models/DisplayWork.scala

```diff
@@ -231,6 +238,14 @@ case object DisplayWork {
           } else None,
     )
 
+  def containsDigitalLocation(items: List[Item[_]]): Boolean =
+    items.exists { item =>
+      item.locations.exists {
+        case _: DigitalLocationDeprecated => true
+        case _                            => false
+      }
+    }
```

Comment:
Might be beyond the scope of this currently, but maybe this field should be calculated upfront during denormalisation? I wanted to make an optimisation to the API code, where we dynamically suppress fields returned from the ES request in relation to the includes, similarily to what we do here:

https://github.com/wellcomecollection/catalogue/blob/b4e4492ef625425b4c4516ed0eba27176504260e/pipeline/relation_embedder/relation_embedder/src/main/scala/uk/ac/wellcome/relation_embedder/RelationsRequestBuilder.scala#L12-L26

This assumes the items exists in the `Work` model, which would not be the case if we did that and had not asked for items

Reply:
Maybe not during "denormalisation", as not really the domain of the `relation_embedder`, but could be generated by transformers and then the merger works out what to do when we merge multiple items.

At the moment we are returning the whole work from ES on every request, including for the works search page which does not need most of that data.

Reply:
I think this is a really good point (although I'm not sure it is in the current scope) and is a great example of the difficulties we have where the internal model is not really what we want the API to be consuming and converting to the display model.

Since this is derived data, from a modelling point of view it probably shouldn't live on the WorkData... If we stopped being tied to storing the internal model in the index, it could just be calculated at index time by the ingestors. I think having to have the merger work out what to do is unnecessary - as I say it's strictly derived data.

Reply:
Not sure I understand why it would be beneficial for this purpose to have a separate model that is stored in the index. Won't we need that field somewhere on the internal model for conversion to the display model?

To me it sounds like the most appropriate place would for it to be in the state of denormalised works, which is the current method of storing derived data.

Reply:
Separating the model that's in the index which gets searched, and the internal model of the pipeline has been something that's come up again and again (and something I know @jamesgorrie is keen on working towards). It would be beneficial here because, whilst it would be good to precalculate this field, that isn't information that's useful as part of the internal model (but there's no other model upon which we can store it).

Reply:
I get why it might be beneficial to have a separate derived model for search, but in this case this information is useful as part of the internal model, precisely as we need this information to be serialised to the display model in the API. The search model as I understand it is data that we do not actually need returned from ES and we just use for indexing purposes.

Reply:
I'm making a larger point that I think that this:
> precisely as we need this information to be serialised to the display model in the API

is not a good thing.

It makes small changes require reindexes because the API is so coupled to the internal model, and it makes storing derived data like this hard.

Unfortunately having thought about it I think you're right - this is too dumb a thing to do on every single work of every single request. I'll add the functionality to the merger but I regard this whole problem as a big to-do that will need refinement at some point.

Reply:
@jamieparkinson I can't say I follow here completely. Are we saying that this property will be derived at the merger stage and then stored in ES?

> It makes small changes require reindexes because the API is so coupled to the internal model, and it makes storing derived data like this hard.

👆 this is a very valid point, and if what I've said above is true, does make this harder.

Reply:
Yes - I think that if we are to calculate it in the pipeline it needs to be done by the merger or the ingestor, and since the ingestor currently is unable to do anything other than index a known model I think it needs to be the merger

Reply:
Cool, yeah I think I understand the problem you are getting at, but not sure I agree that it is an issue with the internal model and API being coupled. One way to solve this using the current modelling would be to have another `WorkState` named `Derived`, where we calculate data like this right at the end of the pipeline. I think when everything is using pipeline storage we can make tools to run a reindex from a specific point in the pipeline (e.g. just recalculate all the derived data rather than having to do the transforms / merging / denormalisation etc before it), which would avoid the whole reindex scenario.

Reply:
> just recalculate all the derived data rather than having to do the transforms / merging / denormalisation etc before it

That's really nice.

>  right at the end of the pipeline

In this case, I think this is what @jamieparkinson is saying, would be the merger as the in_minter / ingestor are intentionally stupid

Reply:
I think `Derived` is a good solution as I don't want to do it in the merger (it just makes the most sense with the current model)

Reply:
FWIW I agree that indexing the internal model is not ideal now and that converting to a more appropriate structure before indexing is a longer term good idea.

Reply:
> converting to a more appropriate structure before indexing is a longer term good idea.

Can someone explain what that structure might look like? I am not sure I understand the issue, as to me it would look exactly the same as the current internal model, but maybe with a few extra fields to help with search and derived fields (in which case a `Work[Derived]` would fit)

Reply:
Separating the pipeline output (internal model) from the API input, and mapping between those 2 data stores separately would make it possible for the API ownership to extend outside of our team and increase the ability of the people working on the website to iterate in the way that they need, rather than being blocked by reindexes, Circe, etc

Reply:
I might be missing something, but still really don't understand how this solves these issues. The suggestion is:

```
pipeline internal model    -->    API internal model         -->    API display model
(pipeline)                        (API backend)                     (API frontend)
```

1) Assuming we are storing a new API model, what is doing this conversion from the pipeline model and storing it, surely its pretty much just an extra pipeline stage? If so, how does that circumvent the need for reindexing?

2) As mentioned before, I do not see in what ways this new API model and the pipeline model (e.g. `Work[Derived]`) would actually differ. Can you elaborate on what the differences might be?

3) The display model is the current interface for people outside the team. There is also no fundamental reason non Scala services cannot use the current internal model if required, or present that as an interface. What purpose does having an extra model serve?

Reply:
I think we're proposing pretty much the same thing if and only if pipeline intermediate storage is used and can be reindexed from any point - that is the same as the 2 data stores I'm talking about. I think any differences are mostly just naming, although it would still be great if that "derived" mapping could be done without having to know/care about the rest of the pipeline (even better if it could be in a language that handles JSON better than Java).

I don't really want this to become another confusing GitHub megathread and it would be good to discuss this in person (well, Teams) with other people (as I say I know @jamesgorrie has thoughts) - maybe later this week?